File: web/src/test/java/no/nav/modiapersonoversikt/service/AnsattServiceImplTest.java
package no.nav.modiapersonoversikt.service;

import _0._0.nav_cons_sak_gosys_3.no.nav.asbo.navansatt.ASBOGOSYSNAVAnsatt;
import _0._0.nav_cons_sak_gosys_3.no.nav.asbo.navansatt.ASBOGOSYSNAVAnsattListe;
import _0._0.nav_cons_sak_gosys_3.no.nav.asbo.navorgenhet.ASBOGOSYSNAVEnhetListe;
import _0._0.nav_cons_sak_gosys_3.no.nav.asbo.navorgenhet.ASBOGOSYSNavEnhet;
import _0._0.nav_cons_sak_gosys_3.no.nav.inf.navansatt.*;
import no.nav.modiapersonoversikt.api.domain.norg.Ansatt;
import no.nav.modiapersonoversikt.api.domain.norg.AnsattEnhet;
import no.nav.modiapersonoversikt.api.utils.http.SubjectHandlerUtil;
import org.junit.jupiter.api.Test;

import java.util.LinkedList;
import java.util.List;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.Is.is;
import static org.mockito.Mockito.*;

class AnsattServiceImplTest {

    private GOSYSNAVansatt ansattWS = mock(GOSYSNAVansatt.class);
    private AnsattServiceImpl ansattServiceImpl = new AnsattServiceImpl(ansattWS);

    @Test
    void skalHenteAlleAnsattaForEnEnhet() throws Exception {
        ASBOGOSYSNAVEnhetListe asbogosysnavEnhetListe = new ASBOGOSYSNAVEnhetListe();
        List<ASBOGOSYSNavEnhet> enhetsliste = asbogosysnavEnhetListe.getNAVEnheter();
        enhetsliste.add(lagNavEnhet("111", "testEnhet"));
        enhetsliste.add(lagNavEnhet("222", "testEnhet2"));
        enhetsliste.add(lagNavEnhet("333", "testEnhet3"));

        when(ansattWS.hentNAVAnsattEnhetListe(any(ASBOGOSYSNAVAnsatt.class))).thenReturn(asbogosysnavEnhetListe);

        List<AnsattEnhet> enheter = SubjectHandlerUtil.withIdent("Z999999", () -> ansattServiceImpl.hentEnhetsliste());
        List<AnsattEnhet> refEnheter = asbogogsysEnhetListeTilAnsattEnhetListe(asbogosysnavEnhetListe);

        verify(ansattWS).hentNAVAnsattEnhetListe(any(ASBOGOSYSNAVAnsatt.class));

        assertThat(enheter.get(0).enhetId, is(refEnheter.get(0).enhetId));
        assertThat(enheter.get(1).enhetId, is(refEnheter.get(1).enhetId));
        assertThat(enheter.get(2).enhetId, is(refEnheter.get(2).enhetId));
    }

    @Test
    void skalKunneHenteNavnAnsatt() throws HentNAVAnsattFaultGOSYSGeneriskfMsg, HentNAVAnsattFaultGOSYSNAVAnsattIkkeFunnetMsg {

        when(ansattWS.hentNAVAnsatt(any(ASBOGOSYSNAVAnsatt.class))).thenReturn(lagNavAnsatt("Kalle", "Karlsson", "111"));

        String navn = ansattServiceImpl.hentAnsattNavn("123");

        assertThat(navn, is("Kalle Karlsson"));
    }

    @Test
    void skalKunneHenteAnsatteForEnhet() throws HentNAVAnsattListeFaultGOSYSGeneriskMsg, HentNAVAnsattListeFaultGOSYSNAVEnhetIkkeFunnetMsg {
        ASBOGOSYSNAVAnsattListe asbogosysnavAnsattListe = new ASBOGOSYSNAVAnsattListe();
        List<ASBOGOSYSNAVAnsatt> ansattListe = asbogosysnavAnsattListe.getNAVAnsatte();
        ansattListe.add(lagNavAnsatt("Kalle", "Karlsson", "111"));
        ansattListe.add(lagNavAnsatt("Klara", "Svensson", "222"));
        ansattListe.add(lagNavAnsatt("Knut", "Larsson", "333"));
        ansattListe.add(lagNavAnsatt("Kristina", "Johansson", "444"));

        when(ansattWS.hentNAVAnsattListe(any(ASBOGOSYSNavEnhet.class))).thenReturn(asbogosysnavAnsattListe);

        List<Ansatt> ansatte = ansattServiceImpl.ansatteForEnhet(new AnsattEnhet("123", "testEnhet"));
        List<Ansatt> refListeAnsatte = asbogogsysAnsatteListeTilAnsatteListe(asbogosysnavAnsattListe);

        verify(ansattWS).hentNAVAnsattListe((any(ASBOGOSYSNavEnhet.class)));
        assertThat(ansatte.get(0).ident, is(refListeAnsatte.get(0).ident));
        assertThat(ansatte.get(1).ident, is(refListeAnsatte.get(1).ident));
        assertThat(ansatte.get(2).ident, is(refListeAnsatte.get(2).ident));

    }

    private ASBOGOSYSNAVAnsatt lagNavAnsatt(String fornavn, String etternavn, String id) {
        ASBOGOSYSNAVAnsatt navAnsatt = new ASBOGOSYSNAVAnsatt();
        navAnsatt.setAnsattNavn(fornavn + " " + etternavn);
        navAnsatt.setFornavn(fornavn);
        navAnsatt.setEtternavn(etternavn);
        navAnsatt.setAnsattId(id);
        return navAnsatt;
    }

    private ASBOGOSYSNavEnhet lagNavEnhet(String enhetsId, String enhetsNavn) {
        ASBOGOSYSNavEnhet navEnhet = new ASBOGOSYSNavEnhet();
        navEnhet.setEnhetsId(enhetsId);
        navEnhet.setEnhetsNavn(enhetsNavn);
        return navEnhet;
    }

    private List<AnsattEnhet> asbogogsysEnhetListeTilAnsattEnhetListe(ASBOGOSYSNAVEnhetListe liste) {
        List<AnsattEnhet> ansattEnhetList = new LinkedList<>();
        for (ASBOGOSYSNavEnhet asbogosysNavEnhet : liste.getNAVEnheter()) {
            ansattEnhetList.add(new AnsattEnhet(asbogosysNavEnhet.getEnhetsId(), asbogosysNavEnhet.getEnhetsNavn()));
        }
        return ansattEnhetList;
    }

    private List<Ansatt> asbogogsysAnsatteListeTilAnsatteListe(ASBOGOSYSNAVAnsattListe asbogosysnavAnsattListe) {
        List<Ansatt> ansattList = new LinkedList<>();
        for (ASBOGOSYSNAVAnsatt ansatt : asbogosysnavAnsattListe.getNAVAnsatte()) {
            ansattList.add(new Ansatt(ansatt.getFornavn(), ansatt.getEtternavn(), ansatt.getAnsattId()));
        }
        return ansattList;
    }

}

File: web/src/test/java/no/nav/modiapersonoversikt/service/ansattservice/AnsattServiceImplTest.kt
package no.nav.modiapersonoversikt.service.ansattservice

import io.mockk.every
import io.mockk.mockk
import no.nav.common.client.nom.NomClient
import no.nav.common.client.nom.VeilederNavn
import no.nav.common.types.identer.EnhetId
import no.nav.common.types.identer.NavIdent
import no.nav.common.utils.fn.UnsafeSupplier
import no.nav.modiapersonoversikt.consumer.norg.NorgApi
import no.nav.modiapersonoversikt.consumer.norg.NorgDomain
import no.nav.modiapersonoversikt.consumer.norg.NorgDomain.EnhetStatus
import no.nav.modiapersonoversikt.service.ansattservice.domain.Ansatt
import no.nav.modiapersonoversikt.service.ansattservice.domain.AnsattEnhet
import no.nav.modiapersonoversikt.service.azure.AzureADService
import no.nav.modiapersonoversikt.testutils.AuthContextTestUtils
import no.nav.personoversikt.common.test.snapshot.SnapshotExtension
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.RegisterExtension

internal class AnsattServiceImplTest {
    private val nomClient: NomClient = mockk()
    private val norgApi: NorgApi = mockk()
    private val azureADService: AzureADService = mockk()
    private val ansattServiceImpl: AnsattServiceImpl = AnsattServiceImpl(norgApi, nomClient, azureADService)

    @JvmField
    @RegisterExtension
    val snapshot = SnapshotExtension()

    @Test
    fun `skal hente alle ansatte for en enhet`() {
        every { azureADService.hentEnheterForVeileder(NavIdent("Z994404")) } returns
            listOf(EnhetId("111"), EnhetId("222"), EnhetId("333"))
        every { norgApi.hentEnheter() } returns
            mapOf(
                EnhetId("111") to NorgDomain.Enhet("111", "testEnhet", EnhetStatus.AKTIV, false),
                EnhetId("222") to NorgDomain.Enhet("222", "testEnhet2", EnhetStatus.AKTIV, false),
                EnhetId("333") to NorgDomain.Enhet("333", "testEnhet3", EnhetStatus.AKTIV, false),
            )

        val enheter: List<AnsattEnhet> =
            AuthContextTestUtils.withIdent(
                "Z994404",
                UnsafeSupplier {
                    ansattServiceImpl.hentEnhetsliste()
                },
            )

        snapshot.assertMatches(enheter)
    }

    @Test
    fun `skal kunne hente navn ansatt`() {
        every { nomClient.finnNavn(listOf(NavIdent("111"))) } returns listOf(lagNavAnsatt("Kalle", "Karlsson", "111"))

        val navn = ansattServiceImpl.hentVeileder(NavIdent("111")).navn

        snapshot.assertMatches(navn)
    }

    @Test
    fun `skal kunne hente liste over ansatt sine fagområder`() {
        every { azureADService.hentTemaerForVeileder(NavIdent("Z994404")) } returns
            listOf("AAP", "DAG")

        val fagomraader = ansattServiceImpl.hentAnsattFagomrader("Z994404")

        snapshot.assertMatches(fagomraader)
    }

    @Test
    fun `skal kunne hente ansatte for enhet`() {
        every { azureADService.hentAnsatteForEnhet(EnhetId("123")) } returns
            listOf(
                Ansatt("Kalle", "Karlsson", "111"),
                Ansatt("Klara", "Svensson", "222"),
                Ansatt("Knut", "Larsson", "333"),
                Ansatt("Kristina", "Johansson", "444"),
            )

        val ansatteListe =
            ansattServiceImpl.ansatteForEnhet(
                AnsattEnhet(
                    "123",
                    "testEnhet",
                ),
            )

        snapshot.assertMatches(ansatteListe)
    }

    private fun lagNavAnsatt(
        fornavn: String,
        etternavn: String,
        id: String,
    ): VeilederNavn =
        VeilederNavn()
            .setFornavn(fornavn)
            .setEtternavn(etternavn)
            .setNavIdent(NavIdent(id))
            .setVisningsNavn("$fornavn $etternavn")
}
